cardano-db-sync/src/Cardano/DbSync/Plugin/Epoch.hs

```diff
@@ -109,71 +109,62 @@ insertBlock
 insertBlock trce (epochNum, blockNo) tipNo = do
 
   mLatestCachedEpoch <- liftIO (readIORef latestCachedEpochVar)
-  chainTipEpoch <- liftIO $ readIORef latestChainTipEpochVar
+  _chainTipEpoch <- liftIO $ readIORef latestChainTipEpochVar
 
-  let blockDiff = unBlockNo tipNo - blockNo
+  let _blockDiff = unBlockNo tipNo - blockNo
 
   let updateEpochAction :: UpdateEpochCmd
-      updateEpochAction = updateEpochNumLogic epochNum mLatestCachedEpoch chainTipEpoch blockDiff
+      updateEpochAction = updateEpochNumLogic epochNum mLatestCachedEpoch
 
   -- Convert commands into actions.
   case updateEpochAction of
-    UpdateEpoch epochNum' -> do
-        -- For debugging! TO BE REMOVED!
-        liftIO . logInfo trce $ "UpdateEpoch!"
-        liftIO . logInfo trce $ "    epochNum = " <> (show epochNum :: Text)
-        liftIO . logInfo trce $ "    mLatestCachedEpoch = " <> (show mLatestCachedEpoch :: Text)
-        liftIO . logInfo trce $ "    chainTipEpoch = " <> (show chainTipEpoch :: Text)
-        liftIO . logInfo trce $ "    blockDiff = " <> (show blockDiff :: Text)
+    UpdateEpoch epochNum' ->
+        updateEpochNumDefault epochNum' trce
+    UpdateEpochs historyEpochNum' epochNum' -> do
+        _ <- updateEpochNumDefault historyEpochNum' trce
         updateEpochNumDefault epochNum' trce
     DoNotUpdate -> pure $ Right ()
 
 -- |Commands for separating the intructions from it's effect.
 data UpdateEpochCmd
   = UpdateEpoch Word64
+  | UpdateEpochs Word64 Word64
   | DoNotUpdate
   deriving (Eq, Show)
 
 -- |Update logic for the epochs.
 updateEpochNumLogic
   :: Word64
   -> Maybe Word64
-  -> Word64
-  -> Word64
   -> UpdateEpochCmd
-updateEpochNumLogic epochNum mLatestCachedEpoch chainTipEpoch blockDiff = do
-
-      let lastCachedEpoch = fromMaybe 0 mLatestCachedEpoch
+updateEpochNumLogic dbEpochNum mNetworkEpochNum = do
 
```

Comment:
There is a problem with only considering the epoch and not the block or slot number.

When syncing, the reason we do not calculate the epoch table entry for every new block is that it is an expensive calculation that drastically slows down sync speeds (by at least 10x). Therefore when syncing we only calculate the epoch entry once for each old epoch.

There is a similar problem when syncing the last epoch when the last epoch is close the end. Say the chain is at epoch N at slot 19000 of a 21600 slot epoch and syncing from scratch. Epochs `0 .. N-1` will be synced quickly and the epoch table entry only calculated once for each epoch, but for the Nth epoch it will calculate the epoch row for every single block in the epoch so this final epoch will be synced much, much slower than the other epochs. AS the cardano chain becomes more heavily used, the calculation of the epoch table entry will become more expensive.

This is why we need to consider the slot within the epoch for this calculation. For the final epoch we should only be calculating the epoch entry when we are with 15 or so (I used 15, but that should probably be reduced) blocks.

Reply:
> When syncing, the reason we do not calculate the epoch table entry for every new block is that it is an expensive calculation that drastically slows down sync speeds (by at least 10x). 

Yes, that is taken into consideration.

> For the final epoch we should only be calculating the epoch entry when we are with 15 or so (I used 15, but that should probably be reduced) blocks.

We can simplify that. In the last epoch, we can update the epoch as long as we are synced up to the tip. The performance overhead will be marginal and the addition of this logic doesn't complicate the conditions that much.

Will push a fix once I confirm it works correctly.